### PR TITLE
BUG: lib: Fix handling of usecols=[] in loadtxt.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
Before this change, because of statements such as `if usecols:`,
`usecols=[]` was treated the same as `usecols=None`, and all the
columns were used.  With this change, `usecols=[]` means "read no
columns", and an empty array is returned.

The error handling is improved: now if `usecols` is given, and the
number of columns in `usecols` does not equal the number of fields
in a given structured dtype, an exception is raised with a message
that explains the problem.  Previously this mismatch would fail
with an incidental IndexError.

### CODE REVIEW

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -914,68 +914,10 @@ def loadtxt(fname, dtype=float, comments='#', delimiter=None,
...
-    if comments is not None:
-        if isinstance(comments, (str, bytes)):
-            comments = [comments]
-        comments = [_decode_line(x) for x in comments]
-        # Compile regex for comments beforehand
-        comments = (re.escape(comment) for comment in comments)
-        regex_comments = re.compile('|'.join(comments))
-
-    if delimiter is not None:
-        delimiter = _decode_line(delimiter)
```
Comment (eric-wieser) [inline numpy/lib/npyio.py:927]: Could you perhaps do the moving of all these lines as a separate PR? It makes it really quite hard to review the bugfix with it merged into this one.

Comment (WarrenWeckesser) [reply numpy/lib/npyio.py:927]: OK, I'll create a maintenance PR for rearranging the code, and then follow up with the bug fix.

Comment (eric-wieser) [reply numpy/lib/npyio.py:927]: Ideal, thanks!

Comment (WarrenWeckesser) [reply numpy/lib/npyio.py:927]: The maintenance PR is https://github.com/numpy/numpy/pull/16633

**Comment by WarrenWeckesser** [comment]
The maintenance PR (https://github.com/numpy/numpy/pull/16633) is merged, and this PR now has just the changes related to fixing the handling of `usecols=[]`.

File: numpy/lib/npyio.py [hunk 2]
```diff
@@ -1071,6 +1071,15 @@ def read_data(chunk_size):
...
     dtype_types, packing = flatten_dtype_internal(dtype)
 
+    if usecols is not None:
+        if len(dtype_types) > 1 and len(usecols) != len(dtype_types):
+            raise ValueError('the number of columns specified in usecols '
+                             f'({len(usecols)}) does not equal the number of '
+                             f'fields in the given dtype ({len(dtype_types)})')
+        if len(usecols) == 0:
+            shp = (0, 0) if ndmin == 2 else (0,)
+            return np.empty(shp, dtype=dtype)
```
Comment (eric-wieser) [inline numpy/lib/npyio.py:1081]: I'd perhaps expect an a `(N, 0)` array (or maybe `(0, N)`?) where `N` is the number of lines in the file.

File: numpy/lib/npyio.py [hunk 3]
```diff
@@ -1071,6 +1071,15 @@ def read_data(chunk_size):
 
     dtype_types, packing = flatten_dtype_internal(dtype)
 
+    if usecols is not None:
+        if len(dtype_types) > 1 and len(usecols) != len(dtype_types):
```
Comment (eric-wieser) [inline numpy/lib/npyio.py:1075]: perhaps clearer as:
```suggestion
        if len(dtype_types) <= 1:
            pass  # this is ok because ???
        elif len(usecols) != len(dtype_types):
```

File: numpy/lib/npyio.py (hunk 2 above)
Comment (WarrenWeckesser) [reply numpy/lib/npyio.py:1081]: I thought about doing that (specifically `(N, 0)`, or `(0, N)` if `unpack` is True).  But then I wondered if that was a foolish consistency.  However, if in fact most people would expect `(N, 0)`, then that's what it should do.

Comment (eric-wieser) [reply numpy/lib/npyio.py:1081]: Is this early exit even needed? What happens if you let the rest of the function run, to avoiding needing a special case?

Comment (seberg) [reply numpy/lib/npyio.py:1081]: I think we should go with `(N, 0)`, ideally it works without any special handling. Speed is not relevant for such a weird corner case anyway.  It is a bit strange, but you could imaging reading `K` columns from multiple files and concatenating them or so, which only makes sense if the shape includes the `N` correctly.

Comment (WarrenWeckesser) [reply numpy/lib/npyio.py:1081]: I'm working on an update to return (N, 0).  I just found a bug in how `loadtxt` handles nested structured dtypes, so I'll try to fix that, and then get back to this PR.

Comment (WarrenWeckesser) [reply numpy/lib/npyio.py:1081]: For future reference, the bug that I encountered is https://github.com/numpy/numpy/issues/16678.

File: numpy/lib/npyio.py (hunk 3 above)
Comment (anirudh2290) [reply numpy/lib/npyio.py:1075]: although i understand that this should be a rare case, maybe makes sense to make it work for len(dtype_types) >= 1 , for one element structured dtypes ?

**Comment by seberg** [comment]
Closing, the error is now (also coming from Warren via `npreadtext` I think):
```
TypeError: If a structured dtype is used, the number of columns in `usecols` must match the effective number of fields. But 2 usecols were given and the number of fields is 3.
```
Although, I guess we could also use `ValueError` as here.
